src/Modal/index.jsx

```diff
@@ -0,0 +1,140 @@
+import Button, { buttonPropTypes } from '../Button';
+
+
+class Modal extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.close = this.close.bind(this);
+    this.handleKeyDown = this.handleKeyDown.bind(this);
+    this.buttons = [];
+
+    // is there a way to avoid depending on the props for the open state?
+    this.state = {
+      open: props.open,
+      focusIndex: 0,
+    };
+  }
+
+  componentDidMount() {
+    this.buttons[0].focus();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.open !== this.props.open) {
+      this.setState({ open: nextProps.open });
+    }
+  }
+
+  componentDidUpdate() {
+    if (this.state.open) {
+      this.buttons[this.state.focusIndex].focus();
+    }
+  }
+
+  close() {
+    this.setState({ open: false });
+    this.props.onClose();
+  }
+
+  // open() {
+  //   this.setState({ open: true });
+  //   this.closeButton.focus();
+  // }
+
+  handleKeyDown(e) {
+    e.preventDefault();
+    if (e.key === 'Escape') {
+      this.close();
+    } else if (e.key === 'Tab') {
+      if (e.shiftKey) {
+        this.setState({
+          focusIndex: ((this.state.focusIndex - 1) + this.buttons.length) % this.buttons.length,
+        });
+      }
+      this.setState({
+        focusIndex: (this.state.focusIndex + 1) % this.buttons.length,
+      });
+    }
+  }
+
+  renderButtons() {
+    return this.props.buttons.map((button, i) => {
+      let buttonElement = button;
+      let buttonProps = button.props;
+
+      if (button.type !== Button) {
+        buttonProps = button;
+      }
+
+      buttonElement = (<Button
+        {...buttonProps}
+        key={i}
+        inputRef={(input) => { this.buttons[i + 1] = input; }}
+        onKeyDown={this.handleKeyDown}
+      />);
+
+      return buttonElement;
+    });
+  }
+
+  render() {
+    return (
+      <div className={classNames(styles.modal, styles.show, { [styles['modal-open']]: this.state.open })}>
+        <div className={styles['modal-dialog']} role="dialog" >
+          <div className={styles['modal-content']}>
+            <div className={styles['modal-header']}>
+              <h5 className={styles['modal-title']}> {this.props.title} </h5>
+              <Button
+                display="&times;"
+                aria-label="Close"
+                buttonType="light"
+                onClick={this.close}
+                inputRef={(input) => { this.buttons[0] = input; }}
+                onKeyDown={this.handleKeyDown}
+              />
+            </div>
+            <div className={styles['modal-body']}>
+              <p>{this.props.body}</p>
+            </div>
+            <div className={styles['modal-footer']}>
+              {this.renderButtons()}
+              <Button
+                display="Close"
+                buttonType="secondary"
+                onClick={this.close}
+                inputRef={(input) => { this.buttons[this.props.buttons.length + 1] = input; }}
```

Comment:
Is this okay to do? What if the user/developer wants buttons to appear dynamically? Will props change accordingly?

Reply:
As a follow on, originally I had attempted to do something like inputRef={(input) => { this.buttons[this.buttons.length + 1] = input; }}. This was having un-intended side-effects, where the buttons array would grow and have null buttons. I think it's because the renderButtons function is executed asynchronously, and therefore, the close button reference may be inserted in to the array before the other buttons have been rendered, causing weird array structure.